### PR TITLE
29253 Refactor info factory

### PR DESCRIPTION
**Pull request by benjamin-confino**
- [x] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [x] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [x] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

#Resolves 29253

This reworks OpenTelemetryInfoFactory which had grown out of control.

- The code for tracking application state and whether we're in app mode or runtime mode has been moved into OpenTelemtryLifecycleManagerImpl
- Instead of pulling in version specific config via OSGi OpenTelemetryInfoFactory is now an abstract class with per-version impls.
- Most importantly of all. Duplicate methods have been merged into one. We no longer have separate but similar methods to create a runtime instance and an app scoped instance.

The goal here is to make this completely transparent with no behaviour changes visible outside the boundary of the open telemetry package. Testing locally only one FAT test reported any difference, and that was only the class name on a debug level trace statement. Far from reaching the threshold of a breaking change. 

----- 

Zero migration statement. There is one change that affects Telemetry 1.0 and 1.1.

Now, if InstantOn is enabled, OpenTelemetry will be no-op until we reach the `restore()` checkpoint phase. From archives on Slack this looks like the correct decision, but I'm setting a reminder to ask Don to confirm this. If not it will be a trivial forward fix to disable the check for instantOn in 1.0 and 1.1

### CODE REVIEW

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
One untriaged failure, com.ibm.ws.wsat.recovery_fat. This bucket does not even use telemetry

**Comment by benjamin-confino** [comment]
#build

File: dev/io.openliberty.microprofile.telemetry.2.0.logging.internal_fat/fat/src/io/openliberty/microprofile/telemetry/logging/internal_fat/TelemetryFFDCTest.java
Comment (Yas-Franco) [inline dev/io.openliberty.microprofile.telemetry.2.0.logging.internal_fat/fat/src/io/openliberty/microprofile/telemetry/logging/internal_fat/TelemetryFFDCTest.java:83]: Do we need a replacement test for this?

**Review by Yas-Franco — DISMISSED** [review]
We may need one test to replace the one that no longer works but other than that, it all looks good to me

File: dev/io.openliberty.microprofile.telemetry.2.0.logging.internal_fat/fat/src/io/openliberty/microprofile/telemetry/logging/internal_fat/TelemetryFFDCTest.java
Comment (benjamin-confino) [reply dev/io.openliberty.microprofile.telemetry.2.0.logging.internal_fat/fat/src/io/openliberty/microprofile/telemetry/logging/internal_fat/TelemetryFFDCTest.java:83]: Prashanth is working to fix the root cause in another PR, at which point we'll re-enable this test.

**Review by Yas-Franco — DISMISSED** [review]

**Comment by benjamin-confino** [comment]
All should be good to go, but I'm waiting for a second review from Thomas Watson because OpenTelemetryLifecycleManagerImpl has instantOn code.

**Comment by benjamin-confino** [comment]
Latest change was rebasing on integration

**Comment by benjamin-confino** [comment]
#build

File: dev/io.openliberty.microprofile.telemetry.internal.common/src/io/openliberty/microprofile/telemetry/internal/common/info/OpenTelemtryLifecycleManagerImpl.java
Comment (tjwatson) [inline dev/io.openliberty.microprofile.telemetry.internal.common/src/io/openliberty/microprofile/telemetry/internal/common/info/OpenTelemtryLifecycleManagerImpl.java]: This is problematic if using `mpTelemetry-1` as it may set `isRuntimeEnabled` to `true` when I don't think it should.  I suggest not registering the hook if using  `mpTelemetry-1`

**Review by tjwatson — CHANGES_REQUESTED** [review]

File: dev/io.openliberty.microprofile.telemetry.internal.common/src/io/openliberty/microprofile/telemetry/internal/common/info/OpenTelemtryLifecycleManagerImpl.java
Comment (benjamin-confino) [reply dev/io.openliberty.microprofile.telemetry.internal.common/src/io/openliberty/microprofile/telemetry/internal/common/info/OpenTelemtryLifecycleManagerImpl.java]: This got me thinking and I double checked some stuff. We actually do have a difference in behaviour for instnatOn in app mode too. But I've made it check `mpTelemetry-1` during the restore too.

**Comment by benjamin-confino** [comment]
#build

File: dev/io.openliberty.microprofile.telemetry.internal.common/src/io/openliberty/microprofile/telemetry/internal/common/info/OpenTelemtryLifecycleManagerImpl.java
Comment (tjwatson) [inline dev/io.openliberty.microprofile.telemetry.internal.common/src/io/openliberty/microprofile/telemetry/internal/common/info/OpenTelemtryLifecycleManagerImpl.java:83]: Since you do the same thing in the checkpoint hook restore method and the case where the hook is not added this is a candidate for using `CheckpointPhase.onRestore`.  Something like this:

```java
waitingForCheckpointRestore = true;
CheckpointPhase.onRestore(() -> {
    waitingForCheckpointRestore = false;
    checkThenSetRuntimeFields(telemetry2OrLater);
});
```

The lambda will run synchronously if not doing a checkpoint or will run on restore for the checkpoint case.

This is just an improvement suggestion.  You can decide to do that later I think.  I'll leave it up to you.

**Review by tjwatson — APPROVED** [review]
I approve the checkpoint related changes.  I have a small improvement that I think makes the code easier to understand, but doesn't block merging.

**Comment by benjamin-confino** [comment]
Build analysis: 

1. `io.openliberty.checkpoint_fat_transaction` - doesn't even use the mpTelemetry feature. 
2. `io.openliberty.microprofile.telemetry.1.0.internal.container_fat` intermittent failure where it cannot find `JvmGcDuration`. Because I can see all the other metrics, and because its intermittent, I know that `mergeInJVMMetrics` is running correctly. Looking at the test I think this because we don't force the GC. Unrelated but must be fixed in another PR
3. There is a 40 second gap in the logs, in the middle of that gap a http request returned zero bytes. Telemetry is on the server but unrelated to the issue

File: dev/io.openliberty.microprofile.telemetry.internal.common/src/io/openliberty/microprofile/telemetry/internal/common/info/OpenTelemtryLifecycleManagerImpl.java
Comment (benjamin-confino) [reply dev/io.openliberty.microprofile.telemetry.internal.common/src/io/openliberty/microprofile/telemetry/internal/common/info/OpenTelemtryLifecycleManagerImpl.java:83]: I like this a lot, it will make a good tool for my arsenal. Since this PR is a blocker for others I'll implement it in a follow up rather than get another build.

**Review by Yas-Franco — APPROVED** [review]

**Pull request merged**
